src/FAI3_backend/src/llm_model.rs

```diff
@@ -0,0 +1,152 @@
+use crate::{
+    check_cycles_before_action, only_admin, LLMModel,
+    ModelDetails, LLM_MODELS, NEXT_MODEL_ID
+};
+use candid::Principal;
+use std::vec;
+use crate::types::{ContextAssociationTestDataPoint, ContextAssociationTestMetricsBag};
+use crate::utils::is_llm_owner;
+
+#[ic_cdk::update]
+pub fn add_llm_model(model_name: String, hf_url: String, model_details: ModelDetails) -> u128 {
+    only_admin();
+    check_cycles_before_action();
+
+    if model_name.trim().is_empty() {
+        ic_cdk::api::trap("Error: LLMModel name cannot be empty or null.");
+    }
+
+    let caller: Principal = ic_cdk::api::caller();
+
+    LLM_MODELS.with(|models| {
+        NEXT_MODEL_ID.with(|id| {
+            let current_id = *id.borrow().get();
+
+            models.borrow_mut().insert(
+                current_id,
+                LLMModel {
+                    model_id: current_id,
+                    model_name,
+                    hf_url,
+                    owners: vec![caller],
+                    cat_metrics: None,
+                    details: model_details,
+                    cat_metrics_history: Vec::new(),
+                },
+            );
+
+            id.borrow_mut().set(current_id + 1).unwrap();
+
+            current_id
+        })
+    })
+}
+
+#[ic_cdk::update]
+pub fn delete_llm_model(model_id: u128) {
+    check_cycles_before_action();
+    let caller: Principal = ic_cdk::api::caller();
+
+    LLM_MODELS.with(|models| {
+        let mut models = models.borrow_mut();
+        let model = models.get(&model_id).expect("LLMModel not found");
+        is_llm_owner(&model, caller);
+        models.remove(&model_id);
+    });
+}
+
+#[ic_cdk::query]
+pub fn get_all_llm_models() -> Vec<LLMModel> {
+    check_cycles_before_action();
+
+    LLM_MODELS.with(|models| {
+        let models = models.borrow();
+        models.values().map(
+            |model| {
+                let mut m2 = model.clone();
+                // keys below are excluded to avoid sending unnecessary information
```

Comment:
I'm removing all the information regarding metrics and data points because it's large. This information can be obtained with the other two methods below this one. But if the interface needs the metrics it can be shown without the data points to avoid an additional call.